### DISCUSSION TITLE
Documentation for writing a new async dialect

### DISCUSSION THREAD

**Discussion by cancan101** [sqlalchemy/sqlalchemy#7854] [Ideas]
It would be great if there was some documentation and / or a spec on what an async dialect looks like.

I see for example that the dialect needs to define:
https://github.com/sqlalchemy/sqlalchemy/blob/7a336f42704a7de214eff560a5b35e272a8050c5/lib/sqlalchemy/engine/interfaces.py#L892-L893

other than that, what methods must be defined _as `async`_ for a new / adapted dialect to be used in an async manner. For reference there is this create doc that outlines a dialect: https://docs.sqlalchemy.org/en/14/core/internals.html#sqlalchemy.engine.Dialect but it does not make clear what must change for async support.

**Comment by zzzeek** [comment]
current paragraph basically says to read the source, we can accept PRs that want to extend on this:  https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst#asyncio-dialects

an asyncio dialect looks just like a sync one, the extra work to do is to create a pep-249 DBAPI adapter for the driver in use.

**Reply by cancan101** [reply]
Is there any documentation on the `server_side` parameter that the cursor method on the `AdaptedConnection`s? [e.g. this](https://github.com/sqlalchemy/sqlalchemy/blob/7f1a3f22abffc1529100e14fcfd07a46a49fd44f/lib/sqlalchemy/dialects/sqlite/aiosqlite.py#L191-L195). The parameter is [not documented on the PEP](https://peps.python.org/pep-0249/#cursor) and I don't really see where `cursor` gets called with anything other than the default value (ie not specified at the call point).

In part I am trying to understand this parameter as the server side implementations are far simpler than the non server side implementations. From what I can tell, the non-server side implementations tend to eagerly pull the data where as the server side ones just proxy through the calls to the underlying cursor. 

In the context of something like sqlite what is the main difference?

*Edit* okay I see about more how they are instantiating https://github.com/sqlalchemy/sqlalchemy/blob/b7ca7ed1b1e9dc67e1c6498a838e03880ad5e8c6/lib/sqlalchemy/engine/default.py#L1256

**Reply by zzzeek** [reply]
the server_side parameter refers to DBAPIs that allow for a non-buffered result of some kind.   in asyncio, this would mean that the database driver provides for an async "fetch" method of some kind.

it's not part of pep-249 as it's an optional thing provided by some (most?) backends, such as psycopg2 provides it via [named cursors](https://www.psycopg.org/docs/usage.html#server-side-cursors).

> In part I am trying to understand this parameter as the server side implementations are far simpler than the non server side implementations. From what I can tell, the non-server side implementations tend to eagerly pull the data where as the server side ones just proxy through the calls to the underlying cursor.

server side implementations are much more complicated in my view.   each fetch of a row implies IO.   so right off, it's inefficient to fetch just one row at a time; you instead need to buffer sets of rows for efficiency.  then, most drivers have a lot of quirks when they deal with server side mode, like not being able to run other statements when you're in the middle of a result set (users frequently want to run queries as they iterate through results) as well as various errors if cursors aren't explicitly closed, errors when trying to get at attributes like "lastrowid", and finally just general incompatibilities, like in PostgreSQL you can't run CREATE TABLE or other kinds of non-row-returning statements with a server side cursor.

within SQLAlchemy's asyncio implementation we have a whole separate result set object that makes use of the async / server side paradigm.

On sqlite, none of this is "real",  since SQLite doesn't use the network, it's all just emulated.
